parsec/data.c

```diff
@@ -565,3 +590,42 @@ parsec_data_destroy( parsec_data_t *data )
 #endif
     PARSEC_OBJ_RELEASE(data);
 }
+
+#include "parsec/utils/debug.h"
+
+int parsec_data_release_self_contained_data(parsec_data_t *data)
+{
+    int32_t nb_copies = data->nb_copies;
+    if (data->super.obj_reference_count != nb_copies) return 0;
+    parsec_data_copy_t *copy = NULL;
+    /* this data is only referenced by it's own copies. If these copies are also only referenced by
+     * data, then we can release them all.
+     */
+    for( uint32_t i = 0; i < parsec_nb_devices; i++) {
+        if (NULL == (copy = data->device_copies[i])) continue;
+        if( copy->super.super.obj_reference_count > 1 || copy->readers > 0 )
+            return 0;
+    }
+    PARSEC_DEBUG_VERBOSE(90, parsec_debug_output, "Release copy %p from self-contained data %p", copy, data);
+    for( uint32_t i = 0; i < parsec_nb_devices; i++) {
+        if (NULL == (copy = data->device_copies[i])) continue;
+        assert(1 == copy->super.super.obj_reference_count && 0 == copy->readers);
+        if( 0 == copy->device_index ) {
```

Comment:
Is this check sufficient to handle the non-GPU devices (like recursive)?

Reply:
updated to use the device_is_gpu function